net: dns: Save info about DHCP when configuring DNS servers

jukkar opened the pull request:
Remember which DNS server was added by a DHCPv4 or v6 message. This will allow system to remove DNS servers that were added by DHCPv4/6. Then when stopping DHCP, we are not leaving DNS servers hanging in the system.

jukkar committed:
net: dns: Save info about source when configuring DNS servers

Remember which DNS server was added by a source like DHCPv4 or v6
message. This will allow system to remove DNS servers that were added by
that source. Then when stopping for example DHCP, we can remove those
specific DNS servers and not leaving DNS servers hanging in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: dhcp: Remove only added DNS servers when stopping

Make sure that we remove only the added DNS servers when
the DHCP is stopped.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: dns: Print info about DHCP added servers

Print information which DNS servers were added by DHCP when
listing DNS servers in "net dns" command. This helps debugging
DNS server issues.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented:
Depending on which PR gets merged first, the tests will need to be adjusted, see #91902

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1942,6 +1941,11 @@ void net_dhcpv4_stop(struct net_if *iface)
 			net_dhcpv4_state_name(iface->config.dhcpv4.state));
 
+		if (IS_ENABLED(CONFIG_NET_DHCPV4_DNS_SERVER_VIA_INTERFACE)) {
+			dns_resolve_remove_via_dhcp(dns_resolve_get_default(),
```
It should not matter much for the interface down, we cannot use the DNS via that interface anyway so calling the `dns_resolve_remove()` should not make much difference here.

jukkar commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -359,6 +359,9 @@ struct dns_resolve_context {
 		uint8_t is_llmnr : 1;
 
+		/** Is this server added via DHCPv4/6 message */
+		uint8_t is_via_dhcp : 1;
```
But then we would not know which one is via DHCP or RA, so when removing DNS address that was added by DHCP we would remove also DNS added by RA which is probably not wanted.

rlubos replied:
So maybe we need to extend this approach, and store an enum value indicating how the server was configured (simliar way we do with IP addresses)?

jukkar commented:
* Updated according to comments. I added enum value so that we can add/remove DNS servers according to the usage.

rlubos commented on include/zephyr/net/dns_resolve.h:
```diff
@@ -526,12 +547,14 @@ int dns_resolve_close(struct dns_resolve_context *ctx);
 			    const char *servers_str[],
-			    const struct sockaddr *servers_sa[]);
+			    const struct sockaddr *servers_sa[],
+			    enum dns_server_source source);
```
Probably worth to mention this change in the migration guide. The API is unstable so we don't need to go any further I think?

jukkar replied:
Sure, I will add that

rlubos reviewed (dismissed):
Nice

jukkar committed:
doc: migration-guide-4.2: Add DNS server reconfiguration info

Add information how the DNS reconfiguration API functions
dns_resolve_reconfigure() and dns_resolve_reconfigure_with_interfaces()
have changed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt reviewed (dismissed):
Couple nits

jukkar commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -488,13 +488,36 @@ static int get_free_slot(struct dns_resolve_context *ctx)
+		return "invalid";
+	}
+
+	return "";
```
That is a good question, I recall that some static analyzers complained if there was no return in a function that was suppose to return something. I think we could remove the default setting as that is probably not really needed.

jukkar commented on subsys/net/lib/dns/resolve.c:
```diff
@@ -488,13 +488,34 @@ static int get_free_slot(struct dns_resolve_context *ctx)
+		return "unknown";
+	case DNS_SOURCE_MANUAL:
+		return "manual";
+	case DNS_SOURCE_DHCPV4:
+		__fallthrough;
```
Hmm, I did not know that. Anyway, putting fallthrough here should do no harm so I would prefer to leave it there.

jukkar commented:
> Hmm, running the test does not fail with this PR, investigating...

Never mind, misunderstood how the test works, will remove extra checks so that we test the DNS server removal.

jukkar committed:
tests: net: dhcpv4: client: Remove checks for DNS server removal

As the net_dhcpv4_stop() will now properly remove DNS servers, remove
the loop check so that the add event gets tested again.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt commented on tests/net/dhcpv4/client/src/main.c:
```diff
@@ -746,17 +746,14 @@ ZTEST(dhcpv4_tests, test_dhcp)
-					      EVT_DNS_SERVER3_ADD,
-				      "Missing DNS server(s) %08x", evt);
-		}
+		evt = k_event_wait_all(&events,
+				       EVT_DNS_SERVER1_ADD | EVT_DNS_SERVER2_ADD |
+				       EVT_DNS_SERVER3_ADD,
+				       false, WAIT_TIME);
```
Should the same check be done for the deletion of the DNS servers after `net_dhcpv4_stop`? I know this test would otherwise fail, but that would make it more explicit.

jukkar committed:
tests: net: dhcpv4: client: Add DNS server deletion test

Make sure net_dhcpv4_stop() will generate DNS server deletion events.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>
